boa/src/syntax/parser/mod.rs

```diff
@@ -83,27 +124,144 @@ impl Parser {
+    fn get_next_token_skip_lineterminator(&mut self) -> Result<Token, ParseError> {
+        self.pos += 1;
+        loop {
+            let token = self.get_token(self.pos)?;
+            if token.kind != TokenKind::LineTerminator {
+                return Ok(token);
+            }
+        }
+    }
+
+    /// Returns the current token  the cursor is sitting on
+    fn get_current_token(&self) -> Result<Token, ParseError> {
+        self.get_token(self.pos)
+    }
+
+    /// Move the cursor back 1
+    fn step_back(&mut self) {
+        self.pos -= 1;
+    }
+
+    /// peeks at the next token
+    fn peek(&self, num: usize) -> Result<Token, ParseError> {
+        self.get_token(self.pos + 1 + num)
+    }
+
+    /// get_current_pos
+    fn get_prev_pos(&self) -> usize {
+        self.pos - 1
+    }
+
+    /// get_current_pos
+    fn get_current_pos(&self) -> usize {
+        self.pos
+    }
+
+    /// set the current position
+    fn set_current_pos(&mut self, pos: usize) -> Result<(), ParseError> {
+        if pos < self.tokens.len() {
+            self.pos = pos;
+            Ok(())
+        } else {
+            Err(ParseError::RangeError)
+        }
+    }
+
+    /// Peek the next token.
+    /// Skipping line terminators.
+    pub fn peek_skip_lineterminator(&mut self) -> Result<Token, ParseError> {
+        let len = self.tokens.len();
+        for i in self.pos..len {
+            let tok = self.tokens[i].clone();
+            if tok.kind != TokenKind::LineTerminator {
+                return Ok(tok);
+            }
+        }
+
+        Err(ParseError::NormalEOF)
+    }
+
+    /// Get the next token.
+    /// Skipping line terminators.
+    pub fn next_skip_lineterminator(&mut self) -> Result<Token, ParseError> {
+        loop {
+            let tok = self.get_next_token()?;
+            if tok.kind != TokenKind::LineTerminator {
+                return Ok(tok);
+            }
+        }
+    }
+
+    /// Peek the next token, and when token is kind:TokenKind, get the token
+    /// Otherwise None
+    /// Skipping line terminators.
+    fn next_if_skip_lineterminator(&mut self, kind: TokenKind) -> Result<bool, ParseError> {
+        match self.peek_skip_lineterminator() {
+            Ok(tok) => {
+                if tok.kind == kind {
+                    self.next_skip_lineterminator()?;
+                    Ok(true)
+                } else {
+                    Ok(false)
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Peek the next token and if it is ``kind``, get the next token, return true.
+    /// Otherwise, return false.
+    fn next_if(&mut self, kind: TokenKind) -> Option<Token> {
+        match self.peek(0) {
+            Ok(tok) => {
+                if tok.kind == kind {
+                    Some(self.get_next_token().unwrap())
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        }
+    }
```

Comment:
Maybe we can implement the `Iterator` trait in the parser? Many of these methods seem pretty similar to the ones used by that trait, and could maybe make it richer.